### PR TITLE
test(tests): add edge_cases instance tests and testing guide

### PR DESCRIPTION
**Pull request by 0xVolosnikov**
## What ❔

Adds Step 4 of the PR544 testing refactor as a reviewable follow-up on top of Step 3:

- introduces new `tests/instances/edge_cases` integration-test crate (16 tests)
- ports edge-case scenarios to `TestingFramework` + `TxBuilder` style
- adds `tests/TESTING.md` with a concise, AI-friendly testing guide
- links the new guide from `tests/README.md`
- wires `tests/instances/edge_cases` into workspace members

## Why ❔

This continues the refactor goal of making tests cleaner, more consistent, and easier to author/review.

Using a single high-level pattern (`TestingFramework`, `TxBuilder`, assertion macros, run-config presets) reduces boilerplate and makes it easier for both humans and AI agents to add/modify tests safely.

## Is this a breaking change?
- [ ] Yes
- [x] No

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted.

## Validation

- `cargo fmt -p edge_cases`
- `cargo check -p edge_cases --quiet`
- `cargo test -p edge_cases --quiet`

### CODE REVIEW

**Comment by 0xVolosnikov** [comment]
[Codex] Architecture review notes:
1) `large_calldata_does_not_panic` currently matches `Ok(_) | Err(_)` and always passes if a result exists, so it does not protect any invariant. Please assert a concrete expected property (for example non-panic + specific success/failure class and no state corruption).
2) `transfer_gas_within_bounds` now uses `assert_gas_used_lt(..., 400_000)`, which is very loose relative to observed transfer costs and provides weak regression signal. Consider tightening to a realistic upper bound (with rationale) so the test remains useful for complexity/cost regressions.

File: tests/instances/edge_cases/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+        .calldata(large_calldata)
+        .gas_limit(5_000_000)
+        .build();
+    let output = tester.execute_block(vec![tx]);
+    match &output.tx_results[0] {
+        Ok(_) | Err(_) => {}
+    }
+}
+
+#[test]
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply tests/instances/edge_cases/src/lib.rs]: [Codex] Fixed in `810c620a`. Replaced the no-op match with explicit assertions (`tx_results.len() == 1` and tx success), so the test now enforces concrete behavior.

File: tests/instances/edge_cases/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+fn transfer_gas_within_bounds() {
+    let signer = PrivateKeySigner::random();
+    let sender = signer.address();
+    let recipient = address!("deadbeef00000000000000000000000000000002");
+
+    let mut tester = new_tester().with_balance(sender, U256::from(DEFAULT_BALANCE));
+    let tx = TxBuilder::new()
+        .from(signer)
+        .to(recipient)
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply tests/instances/edge_cases/src/lib.rs:159]: [Codex] Fixed in `810c620a`. `transfer_gas_within_bounds` now sends non-zero value (`.value(1)`) and uses a tighter gas upper bound (`320_000`).

File: tests/README.md [hunk 3]
```diff
@@ -8,4 +8,8 @@ This directory contains the following subdirectories:
 - `contracts_wasm` - wasm contracts to be used in the tests (currently not used as wasm tests disabled)
 - `forge` - forge project with test solidity contracts
 - `instances` - test cases implemented using rig
-- `fuzzer`
\ No newline at end of file
+- `fuzzer`
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply tests/README.md]: [Codex] Fixed in `810c620a`. Added the missing `fuzzer` description in `tests/README.md` to keep the directory guide consistent.
